parquet/src/main/java/org/apache/iceberg/parquet/TypeWithSchemaVisitor.java

```diff
@@ -19,187 +19,50 @@
-            Preconditions.checkArgument(repeatedKeyValue.getFieldCount() <= 2,
-                "Invalid map: repeated group does not have 2 fields");
-
-            Types.MapType map = null;
-            Types.NestedField keyField = null;
-            Types.NestedField valueField = null;
-            if (iType != null) {
-              map = iType.asMapType();
-              keyField = map.fields().get(0);
-              valueField = map.fields().get(1);
-            }
-
-            visitor.fieldNames.push(repeatedKeyValue.getName());
-            try {
-              T keyResult = null;
-              T valueResult = null;
-              switch (repeatedKeyValue.getFieldCount()) {
-                case 2:
-                  // if there are 2 fields, both key and value are projected
-                  keyResult = visitField(keyField, repeatedKeyValue.getType(0), visitor);
-                  valueResult = visitField(valueField, repeatedKeyValue.getType(1), visitor);
-                  break;
-                case 1:
-                  // if there is just one, use the name to determine what it is
-                  Type keyOrValue = repeatedKeyValue.getType(0);
-                  if (keyOrValue.getName().equalsIgnoreCase("key")) {
-                    keyResult = visitField(keyField, keyOrValue, visitor);
-                    // value result remains null
-                  } else {
-                    valueResult = visitField(valueField, keyOrValue, visitor);
-                    // key result remains null
-                  }
-                  break;
-                default:
-                  // both results will remain null
-              }
-
-              return visitor.map(map, group, keyResult, valueResult);
-
-            } finally {
-              visitor.fieldNames.pop();
-            }
-
-          default:
-        }
-      }
-
-      Types.StructType struct = iType != null ? iType.asStructType() : null;
-      return visitor.struct(struct, group, visitFields(struct, group, visitor));
+  @Override
+  protected Type arrayElementType(Type arrayType) {
+    if (arrayType == null) {
+      return null;
     }
-  }
 
-  private static <T> T visitField(Types.NestedField iField, Type field, TypeWithSchemaVisitor<T> visitor) {
-    visitor.fieldNames.push(field.getName());
-    try {
-      return visit(iField != null ? iField.type() : null, field, visitor);
-    } finally {
-      visitor.fieldNames.pop();
-    }
+    return arrayType.asListType().elementType();
   }
 
-  private static <T> List<T> visitFields(Types.StructType struct, GroupType group, TypeWithSchemaVisitor<T> visitor) {
-    List<T> results = Lists.newArrayListWithExpectedSize(group.getFieldCount());
-    for (Type field : group.getFields()) {
-      int id = -1;
-      if (field.getId() != null) {
-        id = field.getId().intValue();
-      }
-      Types.NestedField iField = (struct != null && id >= 0) ? struct.field(id) : null;
```

Comment:
@rdblue ,  Is it a bug in the master branch here ?   we use the parquet field id to access the iceberg nested field ?  That sounds unreasonable...

Reply:
There are two ways to traverse the schemas together, by name and by ID. When we are building a reader, we use the ID method because names don't necessarily match between when the file was written and the current table schema. The columns are identified by ID, so this is correct in that case.

Traversing two schemas by name is only done when we know that the names between the two match. For example, when Spark runs a CTAS operation, we convert the Spark schema to an Iceberg schema and we know that the two have the same structure and field names, but the Spark schema has no IDs. So when we build a writer, we have to match by name but we know that this is safe. (The two schemas are needed to build writers that convert, like short -> int.)

Reply:
> The columns are identified by ID, so this is correct in that case.

OK, I saw the the [ParquetWriter](https://github.com/apache/iceberg/blob/c28d1c8ba6322af07f0206c4cf3fdad875f37ac1/parquet/src/main/java/org/apache/iceberg/parquet/ParquetWriter.java#L93)  will persist the parquet schema which was converted from iceberg schema by [TypeToMessageType](https://github.com/apache/iceberg/blob/c28d1c8ba6322af07f0206c4cf3fdad875f37ac1/parquet/src/main/java/org/apache/iceberg/parquet/TypeToMessageType.java#L84) (it will use the same field id for the converted parquet field).  So the `TypeWithSchemaVisitor` 's parquet schema should have the same field ids as the iceberg table's schema.  

I just want to make sure it was designed intentionally.  Thanks for the context and confirmation.